### PR TITLE
Update the way DataPostprocessor receives its input.

### LINKED ISSUE #391: Implement a better interface for DataPostprocessor
DataPostprocessor currently declares virtual functions like this:

```
  virtual
  void
  compute_derived_quantities_scalar (const std::vector<double>         &uh,
                                     const std::vector<Tensor<1,dim> > &duh,
                                     const std::vector<Tensor<2,dim> > &dduh,
                                     const std::vector<Point<dim> >    &normals,
                                     const std::vector<Point<dim> >    &evaluation_points,
                                     std::vector<Vector<double> >      &computed_quantities) const;
```

The problem is that if we ever decide that we want to pass more information (e.g., the `material_id` of the cell on which we are currently evaluating, so that a user program can look up cell information), then this doesn't work: we can't just change the signature of the function to pass an additional value, since that would make user programs incompatible. In the past, we added a new virtual function with an extra argument whose default implementation just called the old function with the additional argument dropped. If a user overloads the old function, then it is called; if the user overloads the new function, then that works as well.

Clearly, this does not scale and we will be in trouble every time we want to add an argument. The way I envision this can be done in a reasonably extensible way is to declare them as follows:

```
  struct ScalarInputs {
      const std::vector<double>         uh,
      const std::vector<Tensor<1,dim> > duh,
      const std::vector<Tensor<2,dim> > dduh,
      const std::vector<Point<dim> >    normals,
      const std::vector<Point<dim> >    evaluation_points,
  };

  virtual
  void
  compute_derived_quantities_scalar (const ScalarInputs &scalar_inputs,
                                     std::vector<Vector<double> >      &computed_quantities) const;
```

Using such a scheme, we could add new fields to the `ScalarInputs` structure without breaking old programs -- they'd simply ignore the additional field, but it wouldn't affect the signature of the function and consequently the ability to overload.

### PR DESCRIPTION
**Pull request by bangerth**
In the past, the two functions we use to postprocess data got lists of arguments
for the solution, its derivatives, the evaluation points, and normal vectors.
This is not flexible enough: We can not easily add other information that
we have needed in the past or that users have requested, such as a pointer
to the cell we're currently on, or the material-id of the cell.

Rather than adding each possible argument anyone may want to use
individually to the list of the postprocessor function arguments, we use
a technique here that has proven useful in ASPECT: the functions simply
take a reference to a structure that contains these individual pieces
of information. We can extend the members of these structures without
backward compatibility issues because the functions still get a
reference to the same structure, we just grow the structure
itself. Functions that never used the new members of the structure
will continue to work as they always did.

This at long last fixes #391.

### CODE REVIEW

**Comment by drwells** [comment]
This looks good. My only substantive complaint is that the variable names could be improved: would you mind using `values`, `gradients`, and `hessians` (or similar) instead of `uh`, `duh`, and `dduh`? I find the current variables hard to read.

**Comment by bangerth** [comment]
Fair point -- I renamed these variables to `solution_values/gradients/hessians`.

File: tests/numerics/data_out_postprocessor_01-new.cc [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+  void
+  compute_derived_quantities_vector (const DataPostprocessorInputs::Vector<dim> &input_data,
+                                     std::vector<Vector<double> >                    &computed_quantities) const
+  {
+    for (unsigned int q=0; q<input_data.uh.size(); ++q)
+      {
+        AssertThrow (computed_quantities[q].size() == 1,
+                     ExcInternalError());
+
+        computed_quantities[q](0) = input_data.uh[q](0)*input_data.uh[q](0) + input_data.uh[q](1)*input_data.uh[q](1);
```
Comment (drwells) [inline tests/numerics/data_out_postprocessor_01-new.cc]: should be `.solution_values`

File: tests/numerics/data_out_postprocessor_scalar_01-new.cc [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    :
+    DataPostprocessorScalar<dim> ("magnitude", update_values)
+  {}
+
+  virtual
+  void
+  compute_derived_quantities_vector (const DataPostprocessorInputs::Vector<dim> &input_data,
+                                     std::vector<Vector<double> >                    &computed_quantities) const
+  {
+    for (unsigned int q=0; q<input_data.uh.size(); ++q)
```
Comment (drwells) [inline tests/numerics/data_out_postprocessor_scalar_01-new.cc]: Same

**Comment by bangerth** [comment]
I did that for all tests I could find as well.

I'll take care of corresponding changes in `examples/` in a separate patch later.

**Comment by drwells** [comment]
Looks good to me. It would be nice to squash things down but if you don't want to then please merge this yourself.

**Comment by bangerth** [comment]
I squashed the renaming into the commits that change the functionality and add tests, respectively.

**Pull request merged**